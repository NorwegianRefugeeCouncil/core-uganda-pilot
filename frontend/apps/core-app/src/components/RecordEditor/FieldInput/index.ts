export { FieldInput } from './FieldInput.container';
