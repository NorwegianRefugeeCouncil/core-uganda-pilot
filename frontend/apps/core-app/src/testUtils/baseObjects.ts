import { FormDefinition, FormType, Record } from 'core-api-client';

export const baseForm: FormDefinition = {
  id: 'form1',
  code: '',
  name: 'form 1',
  databaseId: 'databaseId1',
  folderId: 'folderId1',
  formType: FormType.DefaultFormType,
  fields: [],
};

export const baseTextField = {
  id: 'field1',
  name: 'field 1',
  description: 'description 1',
  code: '',
  required: false,
  key: false,
  fieldType: {
    text: {},
  },
};

export const baseMultilineTextField = {
  id: 'field2',
  name: 'field 2',
  description: 'description 2',
  code: '',
  required: true,
  key: true,
  fieldType: {
    multilineText: {},
  },
};

export const baseRecord1: Record = {
  id: 'record1',
  values: [
    {
      value: 'text',
      fieldId: 'field1',
    },
    {
      value: 'multi line text',
      fieldId: 'field2',
    },
  ],
  formId: 'form1',
  databaseId: 'databaseId1',
  ownerId: undefined,
};
