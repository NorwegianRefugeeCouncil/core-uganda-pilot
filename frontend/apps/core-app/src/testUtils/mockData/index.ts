export { makeRecord } from './Record';
export { makeForm } from './Form';
export { makeField } from './FormField';
