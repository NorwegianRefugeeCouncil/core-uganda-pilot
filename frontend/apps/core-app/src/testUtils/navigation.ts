export const mockNavigationProp = {
  dispatch: jest.fn(),
  navigate: jest.fn(),
  reset: jest.fn(),
  goBack: jest.fn(),
  isFocused: jest.fn(),
  canGoBack: jest.fn(),
  getParent: jest.fn(),
  getState: jest.fn(),
  setParams: jest.fn(),
  setOptions: jest.fn(),
  addListener: jest.fn(),
  removeListener: jest.fn(),
  replace: jest.fn(),
  push: jest.fn(),
  pop: jest.fn(),
  popToTop: jest.fn(),
};
