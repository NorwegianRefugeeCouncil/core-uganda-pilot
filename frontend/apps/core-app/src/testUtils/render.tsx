import * as React from 'react';
import { NavigationContainer } from '@react-navigation/native';
import {
  render as r,
  RenderAPI,
  RenderOptions,
} from '@testing-library/react-native';
import { render as rWeb } from '@testing-library/react';
import { NativeBaseProvider } from 'native-base';
import { theme } from 'core-design-system';

const inset = {
  frame: { x: 0, y: 0, width: 0, height: 0 },
  insets: { top: 0, left: 0, right: 0, bottom: 0 },
};

export const render = (
  component: React.ReactElement<any>,
  options?: RenderOptions | undefined,
): RenderAPI =>
  r(
    <NativeBaseProvider theme={theme} initialWindowMetrics={inset}>
      <NavigationContainer>{component}</NavigationContainer>
    </NativeBaseProvider>,
    options,
  );

export const renderWeb = (
  component: React.ReactElement<any>,
  options?: RenderOptions | undefined,
) =>
  rWeb(
    <NativeBaseProvider theme={theme} initialWindowMetrics={inset}>
      <NavigationContainer>{component}</NavigationContainer>
    </NativeBaseProvider>,
    options,
  );
