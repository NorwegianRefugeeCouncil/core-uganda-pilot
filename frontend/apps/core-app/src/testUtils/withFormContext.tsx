import { Button } from 'native-base';
import React from 'react';
import { FieldValues, FormProvider, useForm } from 'react-hook-form';

export const withFormContext =
  <T,>(
    Component: React.FC<T>,
    defaultValues: FieldValues = {},
    onSubmit: (data: any) => void = () => {},
  ) =>
  // eslint-disable-next-line react/display-name
  (props: T) => {
    const f = useForm({ defaultValues });

    const handleSubmit = f.handleSubmit((data: any) => {
      onSubmit(data);
    });

    return (
      <FormProvider {...f}>
        <Component {...props} />
        <Button testID="with-form-context-submit-button" onPress={handleSubmit}>
          Submit
        </Button>
      </FormProvider>
    );
  };
