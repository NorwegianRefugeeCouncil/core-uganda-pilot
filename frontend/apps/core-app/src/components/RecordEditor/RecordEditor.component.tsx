import * as React from 'react';
import { Box, VStack, useBreakpointValue } from 'native-base';
import { FieldKind, FormDefinition, getFieldKind } from 'core-api-client';

import { FieldInput } from './FieldInput';

type Props = {
  form: FormDefinition;
};

const useGetFieldWith = () => {
  const width = useBreakpointValue({
    sm: '100%',
    md: '80%',
  });

  return (field: FormDefinition['fields'][number]) => {
    const kind = getFieldKind(field.fieldType);

    if (kind === FieldKind.SubForm) {
      return '100%';
    }

    return width;
  };
};

export const RecordEditorComponent: React.FC<Props> = ({ form }) => {
  const getFieldWidth = useGetFieldWith();

  return (
    <VStack space={4} flexWrap="wrap">
      {form.fields.map((field) => {
        return (
          <Box key={field.id} width={getFieldWidth(field)}>
            <FieldInput form={form} field={field} />
          </Box>
        );
      })}
    </VStack>
  );
};
