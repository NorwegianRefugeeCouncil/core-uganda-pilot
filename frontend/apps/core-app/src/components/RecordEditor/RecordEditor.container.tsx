import * as React from 'react';
import { FormDefinition } from 'core-api-client';
import { useFormContext } from 'react-hook-form';

import { RecordEditorComponent } from './RecordEditor.component';

type Props = {
  form: FormDefinition;
};

export const RecordEditorContainer: React.FC<Props> = ({ form }) => {
  const formContext = useFormContext();

  if (!formContext) throw new Error('Form context is not available');

  return <RecordEditorComponent form={form} />;
};
