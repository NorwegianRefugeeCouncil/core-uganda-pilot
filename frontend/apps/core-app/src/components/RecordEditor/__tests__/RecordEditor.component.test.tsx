import { FormDefinition, FormType } from 'core-api-client';

import { render } from '../../../testUtils/render';
import { RecordEditorComponent } from '../RecordEditor.component';

jest.mock('../FieldInput', () => {
  const { View, Text } = jest.requireActual('react-native');
  return {
    FieldInput: ({
      form,
      field,
    }: {
      form: FormDefinition;
      field: FormDefinition['fields'][number];
    }) => (
      <View>
        <Text>{JSON.stringify(form)}</Text>
        <Text>{JSON.stringify(field)}</Text>
      </View>
    ),
  };
});

it('should match the snapshot', () => {
  const form: FormDefinition = {
    id: 'form-1',
    code: '',
    name: 'form 1',
    databaseId: 'database-id-1',
    folderId: 'folder-id-1',
    formType: FormType.DefaultFormType,
    fields: [
      {
        id: 'field-1',
        name: 'field 1',
        description: 'description 1',
        code: '',
        required: false,
        key: false,
        fieldType: {
          text: {},
        },
      },
      {
        id: 'field-2',
        name: 'field 2',
        description: 'description 2',
        code: '',
        required: true,
        key: true,
        fieldType: {
          multilineText: {},
        },
      },
    ],
  };

  const { toJSON } = render(<RecordEditorComponent form={form} />);
  expect(toJSON()).toMatchSnapshot();
});
