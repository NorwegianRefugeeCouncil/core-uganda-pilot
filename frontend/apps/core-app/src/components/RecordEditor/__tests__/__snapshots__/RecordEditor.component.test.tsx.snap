// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    dataSet={Object {}}
    style={
      Object {
        "flexWrap": "wrap",
      }
    }
  >
    <View
      dataSet={Object {}}
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <View>
        <Text>
          {"id":"form-1","code":"","name":"form 1","databaseId":"database-id-1","folderId":"folder-id-1","formType":"default","fields":[{"id":"field-1","name":"field 1","description":"description 1","code":"","required":false,"key":false,"fieldType":{"text":{}}},{"id":"field-2","name":"field 2","description":"description 2","code":"","required":true,"key":true,"fieldType":{"multilineText":{}}}]}
        </Text>
        <Text>
          {"id":"field-1","name":"field 1","description":"description 1","code":"","required":false,"key":false,"fieldType":{"text":{}}}
        </Text>
      </View>
    </View>
    <View
      dataSet={Object {}}
      style={
        Object {
          "height": 16,
        }
      }
    />
    <View
      dataSet={Object {}}
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <View>
        <Text>
          {"id":"form-1","code":"","name":"form 1","databaseId":"database-id-1","folderId":"folder-id-1","formType":"default","fields":[{"id":"field-1","name":"field 1","description":"description 1","code":"","required":false,"key":false,"fieldType":{"text":{}}},{"id":"field-2","name":"field 2","description":"description 2","code":"","required":true,"key":true,"fieldType":{"multilineText":{}}}]}
        </Text>
        <Text>
          {"id":"field-2","name":"field 2","description":"description 2","code":"","required":true,"key":true,"fieldType":{"multilineText":{}}}
        </Text>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
